File: Lecture21-CommandPattern/Typescript/src/Sequence.ts
import { ISequence } from './ISequence';

export class Sequence implements ISequence {

  private sequence: string = '';

  constructor(sequence?: string) {
    if (sequence) this.sequence = sequence;
  }

  print(): void {
    console.log(this.toString());
  }
  insert(pos: number, s: string): void {
    this.sequence = this.sequence.substr(0, pos) + s + this.sequence.substr(pos);
  }
  delete(pos: number, count: number): void {
    let s = this.sequence.split('');
    s.splice(pos, count);
    this.sequence = s.join('');
  }
  clear(): void {
    this.sequence = '';
  }

  toString(): string {
    return this.sequence.toString();
  }
}

File: Lecture21-CommandPattern/Java/Sequence.java
public class Sequence implements ISequence {
    private StringBuilder _sequence;

    Sequence() {
        _sequence = new StringBuilder();
    }

    @Override
    public void print() {
        System.out.println(this.toString());
    }

    @Override
    public void insert(int pos, String s) {
        _sequence.insert(pos, s);
    }

    @Override
    public void delete(int pos, int count) throws SequenceException {
        try {
            _sequence.delete(pos, pos + count);
        } catch (Exception e) {
            throw new SequenceException();
        }
    }

    @Override
    public void clear() {
        _sequence = new StringBuilder();
    }

    @Override
    public String toString() {
        return _sequence.toString();
    }
}
